### ISSUE TITLE
Getting a "disk I/O error" when running "test_no_conflicts_enabled_with_revs_limit" repeatedly

### ISSUE THREAD

**Issue by hrajput89** [couchbase/couchbase-lite-core#381]
We are testing "no conflict" enabled with revision_limit set test, for multiple options, On running the test with different parameters individually, all test passes. However, when running them in succession to each other, test starts to fail after first successful run. Below are the details of the test - 

>         1. Enable allow_conflicts = false in SG config and specified revs_limit
>         2. Add a few docs through CBL.
>         3. Update the doc few times in CBL.
>         4. Do push replication to Sg with continous replication
>         5. Update docs in CBL again few more times which can cross the revs_limit
>         6. Check the revision list for the doc

The test environment - 
CBL app -
 -  android 2.0.0-291

SG - 
 - OS: Centos7
 - Build - 2.0.0-749
 - mode - cc

CBS -
 - OS: Centos7
 - Build - 5.0.0-3519
 - mode

Test method is available here - https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/testsuites/CBLTester/CBL_Functional_tests/test_no_conflicts_cbl.py#L96

[sgcollect_info.zip](https://github.com/couchbase/couchbase-lite-core/files/1674926/sgcollect_info.zip)
[CBL_app.log](https://github.com/couchbase/couchbase-lite-core/files/1674928/CBL_app.log)
[CBL_testrunner.log](https://github.com/couchbase/couchbase-lite-core/files/1674929/CBL_testrunner.log)

**Comment by snej** [comment]
There's an unexpected warning just above the I/O error log: `W/LiteCore: SQLite error (code 28): file unlinked while open: /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test.cblite2/db.sqlite3`. Deleting an open SQLite database can certainly cause data corruption, but your test description doesn't say anything about deleting the DB.

**Comment by snej** [comment]
@hrajput89 Is it possible you're deleting the database on one thread, and opening it on another before the deletion has finished? I found a race condition in LiteCore that could allow this to happen. (But even if I fix that, it's still a bad idea for you to do a simultaneous delete and open. Those should be serialized.)

**Comment by djpongh** [comment]
@snej , is the race condition fix, checked in? I want to see if QE can re-run with the latest LiteCore.

**Comment by snej** [comment]
It's on the dev branch. I don't know if this is the problem with this test, though. I was hoping for an answer to my question above.

(Even if that is the problem with the test, my commit won't make the test work. It'll just throw an exception.)

**Comment by snej** [comment]
I've been digging through the test suite code, and I think part of the problem is that some of the tests use the RPC function `database_deleteDBbyName` instead of telling the Database instance to delete (`database_deleteDB`). That's a problem because the underlying LiteCore call (`c4db_deleteAtPath`) doesn't coordinate with any open Database objects: it just nukes the files. So if any Database instances are open, the results aren't good. `c4db_deleteAtPath` is meant to be used only when you don't have an open C4Database.

**Comment by snej** [comment]
I should make `c4db_deleteAtPath` less dangerous, but the correct behavior is to have it fail if there is an open database at that path, and that will cause the tests to fail. So the tests should be fixed.

**Comment by hrajput89** [comment]
@snej, The test opens a db before the start of the test and delete after test finishes off. Test is parameterized and follows setup (create a db instance) and teardown (delete the db) model, so db opens before every test and closes after it finishes. For safe side, I added sleep of 10 sec after we delete db, but I'm still seeing the same error.

[no_conflict_test.log](https://github.com/couchbase/couchbase-lite-core/files/1676971/no_conflict_test.log)

**Comment by snej** [comment]
Does this only happen on Android?
The Android logs don't have timestamps, or any indication of what thread is logging, which would help. iOS logs do, so if the problem happens on iOS I'd like to see those logs.

**Comment by hideki** [comment]
Hi @hrajput89  and @sridevi-15,

I have a request and some quetions

- Could you please help us to set up the test environment on our dev machine?
- Is this error reproducible with single test of `test_no_conflicts_enabled_with_revs_limit()`? Or is this reproducible by executing test suite?
- What is test device info? Android version, device model, etc.

NOTE: I have seen the error with code: 19 with unit test in case executing unit test suite (no single test) with Android version 23 or lower on an Android ARM emulator. Following is my assumption, as Database.close() does not wait till associated replicator stops, I assume static method `Database.delete(<db name>)` or `File.delete()` delete the database file even though replicator is still accessing the database.

**Comment by borrrden** [comment]
This theory sounds very plausible.  I've been fighting with disk I/O errors happening once every 50 - 100 runs or so and the test case also behaves in a similar way (c4db_deleteAtPath in the teardown).

**Comment by hrajput89** [comment]
@hideki,

This is the test environment, I'm using.

Genymotion Android emulator Pixel XL 7.1.0 API25
 
SG 
 - build - 2.0.0-749
 - OS - Centos7
 - mode cc

CBS
 - build 5.0.0-3519
 - OS - Centos7

The error is not reproducible when running a single test, but only when we run test suite. 
One can run test using the following command -

pytest  -s  --mode=cc --server-version=5.0.0-3519 --sync-gateway-version=2.0.0-749 --liteserv-platform=android --liteserv-host=192.168.0.126  --liteserv-port=8080  --xattrs  --create-db-per-suite="test_db" --create-db-per-test=pull-test  --no-conflicts -k test_no_conflicts_enabled_with_revs_limit testsuites/CBLTester/CBL_Functional_tests/

**Comment by snej** [comment]
>I assume static method Database.delete(<db name>) or File.delete() delete the database file even though replicator is still accessing the database.

That would happen if you're _not_ calling `c4db_deleteAtPath`. That call will fail with `kC4ErrorBusy` if any C4Database is open on that file, so it's not likely to cause this problem. (I did just fix a race condition, but it's pretty unlikely.) But if you're using File.delete(), then yes, it will cause serious problems if another connection is open.

**Comment by snej** [comment]
@hrajput89, please answer my question "Does this only happen on Android?" If so, it's likely to be an issue in Couchbase Lite, not LiteCore.

**Comment by sridevi-15** [comment]
I see same error on another test too. Here are logs : 
Test code for this test is https://github.com/couchbaselabs/mobile-testkit/blob/feature/cbl20-ver1/testsuites/CBLTester/CBL_Functional_tests/test_no_conflicts_cbl.py#L696 
https://gist.github.com/sridevi-15/c4a4a133597822b85192673b02f8b5b8

Steps to reproduce : 
1. Create Db1, db2, db3 on CBL.
2. create docs on db1, db2, db3,
3. Start replication with continous true and push pull replication from CBL db1, CBL db2,
4. Start replication with continous true and push pull replication from CBL db1, CBL db3
5. wait for replication to be idle for step 3 and step 4 and stop the replicator 
6. Start new replication with CBL db3 to sync-gateway

REsult : 
1. Check the status of replicator to verify replicator started, it shows replicator stopped and fail the test.

**Comment by sridevi-15** [comment]
Note : I have also added debug statement after step 5 to verify cbl db3 exists. it displays that cbl db 3 exists, but replicator fails to start after step 6

**Comment by borrrden** [comment]
Additional info as this just happened again.  Sync Gateway is not required for this scenario.  

> file unlinked while open

This is only logged in one place in the sqlite.c code base, during the `unixClose` method.  It does so because a call to `fstat` indicated 0 hard links to the db file (i.e. it was deleted).  Renaming it would not cause this error, as that case is handled separately (not that renaming was a particularly plausible theory anyway...).  

There doesn't appear to be anywhere that could be calling any delete DB methods at the point in time that this is happening.  For me, it happens in the middle of a test during which no deletes take place (the previous delete would have taken place during the previous test tear down, but if that were affecting things this error would have manifested at the BEGINNING of the test, and not the middle).  It happens *after* the replicator has completely stopped, and called its status changed callback indicating the stoppage, but before the next iteration of the test loop where another replicator is started.  This suggests to me that somehow the file got deleted somehow while an active `C4Database` was using it.

> LiteCore SQLite error (code 522): disk I/O error

This showed up in the second logs, as well as mine that just happened.  This indicates `SQLITE_IOERR_SHORT_READ` which means that SQLite tried to read a certain amount of bytes from a file but the return value of the read was 0 <= amountReceived < amountRequested.  The function that returned it was the C system call `ssize_t pread(int __fd, void * __buf, size_t __nbyte, off_t __offset);`  I'm not sure why this function would return a value in that range instead of -1 indicating failure.  

This only happens at best once in about 200 runs for me (160 disregarding Windows) so if this happens more often for the functional test suite then we might be able to learn more from bumping up the database logging level (or maybe it will make the problem go away by mistake...)

**Comment by hrajput89** [comment]
@snej, we saw this issue in android as if now. We didn't observe this error in iOS and .net is yet to be tested.

Regarding the prefix on logs. This is the format <date and time> <PID-TID> <package name> <tag> <msg>. Hope this helps

**Comment by djpongh** [comment]
Also see: https://github.com/couchbase/couchbase-lite-android/issues/1559

**Comment by sridevi-15** [comment]
i see this often with this scenario : 
1. Create CBL db, create documents.
2. Configure replication with continuous true and with push_pull
3. start replication
4. wait for replication to complete.
5. Verify sync-gateway got replicated. 
6. delete CBL db
7. Create the CBL db with same db name.
8. Repeat steps 2, 3, 4, 
9. Verify replication is completed and verify docs got replicated to sync-gateway

Result : 
docs fail to replicate to sync-gateway and do not see any docs in sync-gateway 

Logs : https://gist.github.com/sridevi-15/7f7a8e0743219a6cc4dbfd1cc6bcc00d

**Comment by sridevi-15** [comment]
I changed the tests to use new database everytime. I will keep running tests and see tests are going fine.
Just an additional log info I see while deleting the database 
02-02 22:27:55.859 8275-8338/? I/LiteCore [SQL]: Created prerecorded query enum with 10 rows (164 bytes) in 0.034ms
02-02 22:27:55.865 8275-8338/? I/Database: Deleting Database@3498239{name='pull-test1517642725.85'} at path /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test1517642725.85.cblite2/
02-02 22:27:55.865 8275-8338/? I/LiteCore [DB]: Closing DataFile
02-02 22:27:55.866 8275-8338/? I/System.out: database delete 
**02-02 22:27:55.874 8275-8284/? W/LiteCore: SQLite error (code 28): file unlinked while open: /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test1517642725.85.cblite2/db.sqlite3
02-02 22:32:14.370 8275-8275/? I/Choreographer: Skipped 239 frames!  The application may be doing too much work on its main thread.**
02-02 22:32:24.024 8275-8282/? I/art: Debugger is no longer active
02-02 22:32:24.024 8275-8282/? I/art: Starting a blocking GC Instrumentation

**Comment by snej** [comment]
>i see this often with this scenario : [...]

We really need to keep Github issues specific to one issue. If you see something unrelated that's worth noting, please file a new issue.

>02-02 22:27:55.874 8275-8284/? W/LiteCore: SQLite error (code 28): file unlinked while open: 

Hmm, this is still the same underlying problem that caused the original bug.

**Comment by snej** [comment]
We need to add a line to `SQLiteDataFile::deleteFile()` that logs the path of the file being deleted. Like so:
```c++
    bool SQLiteDataFile::Factory::deleteFile(const FilePath &path, const Options*) {
        LogTo(DBLog, "Deleting database file %s (with -wal and -shm)", path.path().c_str());
        ...
```
This has already been done on the dev branch, but we need that log in the build being run for the Android DB22 tests.

Once we have that, we can look at the full logs from a run where the error 28 occurs and see what databases are open on the path being deleted.

**Comment by snej** [comment]
@hideki, since this change has to be made to the LiteCore revision currently used by Android DB22, and involves updating the submodule, could you take care of it? Just make a temporary branch there and add that one log call.

**Comment by borrrden** [comment]
> Hmm, this is still the same underlying problem that caused the original bug.

For what it's worth, all these scenarios seem to be ways to trigger the same behavior and so seem relevant to the discussion.

**Comment by hideki** [comment]
Once @sridevi-15 finished test env on mac mini, I will run the test with file delete logging.

**Comment by borrrden** [comment]
@snej Update with the new log statement, there doesn't appear to be any discrepancy between the open and delete paths:

> Opening DataFile /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test.cblite2/db.sqlite3
> Deleting database file /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test.cblite2/db.sqlite3 (with -wal and -shm)

These two statements appeared during the test suite, and a failure was still observed.

**Comment by borrrden** [comment]
Extra piece of maybe relevant information:  The error happens between the time when the replicator tries to open the DB, and this statement:

> {Repl#1}==> N8litecore4repl10ReplicatorE /data/user/0/com.example.hemant.coucbaselitesampleapp/files/pull-test.cblite2/ ->ws:192.168.33.11:4984/db/_blipsync

There were only two instances of this log entry in the logs (for the two passing variants, the failed one did not show this)

**Comment by snej** [comment]
>The warning about file unlinked while open comes during a finalizer call

Interesting. IIRC you can't close a SQLite database while statements are still open; you get a SQLITE_BUSY error. I've run into this in the past when internal LiteCore code wasn't cleaning up statements it created. So this implies either

* The SQLiteDataFile was not closed yet, but somehow a call to `c4db_delete` succeeded (i.e. the race conditions I fixed)
* The SQLiteDataFile was closed but somehow the failure in `sqlite3_close` was ignored.

**Comment by snej** [comment]
I've got it! The destructor of `SQLite::Database` calls `sqlite3_close_v2`, whose behavior is different than the usual close:

>If sqlite3_close_v2() is called on a database connection that still has outstanding prepared statements … then it returns SQLITE_OK and the deallocation of resources is deferred until all prepared statements … are also destroyed.

So the close succeeds, but behind the scenes the SQLite database connection is still open, waiting for the statements to be finalized. Then we delete the files, since we think all the connections are closed. When the finalizer runs (on some background thread) and finalizes the statement, the SQLite connection finally [sic] closes the db but discovers that the file has been unlinked and returns that error.

I _think_ this means that the error is harmless and we can just add some code to suppress logging it.

**Comment by snej** [comment]
PS: There's a warning in the Database destructor, `SQLite::Database %p close deferred due to open sqlite_stmt`, that would have made this much easier to figure out if we'd seen it. Unfortunately it's only being logged to stderr, so it doesn't get captured in the log output, at least on Android. (It doesn't call `Log()` because it's in the SQLiteCpp submodule.)

**Comment by snej** [comment]
What about the `LiteCore SQLite error (code 522): disk I/O error` warning? Is that still happening? I'm pretty sure that also comes from closing the sqlite3_stmt after the database is closed & deleted, but I'm not sure why there would be a "short read".

**Comment by borrrden** [comment]
I came to the same conclusion about the first warning, but the second one is the one actually causing problems.  It's not coming from that same location, but from a query against a non-existent database file during the `c4db_openAgain` call when making a new replicator.  This causes replication failure and stoppage.

**Comment by borrrden** [comment]
To be clearer, the "first warning" in this case is the "unlinked while opened" which is error code 28, and not technically an error.  The "second warning" is actually an error (522) and is caused by the decrypt call when the replicator opens a database to find it inexplicably gone.

**Comment by snej** [comment]
Could you post a backtrace?

**Comment by borrrden** [comment]
I don't know how to copy backtraces from Android Studio and neither does Hideki, unfortunately. I'll try to type up the cliff notes for it though.

**Comment by borrrden** [comment]
c4repl_new                                                   c4Replicator.cc : 113
    c4db_openAgain                                          c4Database.cc : 115
    c4db_open                                                   c4Database.cc : 98
    ....
    litecore::SQLiteDataFile::SQLiteDataFile    SQLiteDataFile.cc : 168
    litecore::SQLiteDataFile::reopen                 SQLiteDataFile.cc : 186
    litecore::SQLiteDataFile::decrypt                SQLiteDataFile.cc : 287
    litecore::SQLiteDataFile::_exec                   SQLiteDataFile.cc : 383

This corresponds to the sanity check at the end of `decrypt()`

```c++
// Verify that encryption key is correct (or db is unencrypted, if no key given):
 _exec("SELECT count(*) FROM sqlite_master");
```

At this point db.sqlite3, db.sqlite3-shm, db.sqlite3-wal are always in various states of missing.  Usually they are all deleted, sometimes just shm and wal are deleted, but the cblite2 folder itself never seems to be deleted.  That suggests that something deleted it between the call to `sqlite3_open` and this query, but I still can't figure out what that is.  It doesn't appear to be anything from the C4 API.

**Comment by snej** [comment]
c4db_openAgain just calls c4db_open with the current db's path and config. So the DataFile::Shared's refcount must be at least 1 already, and will be bumped by one on return, so there's no window where the files could be deleted by LiteCore.

Is there a specific test this occurs with? If so, what's it doing? Could you post the logs, with DB turned up to Debug?

**Comment by borrrden** [comment]
The test that is doing this is the one in the title of this ticket "test_no_conflicts_enabled_with_revs_limit" in the functional test suite.  The details of the test are in the first comment on the ticket, though I don't know if they are detailed enough.  The key here is that this test is run three times with differing parameters (number of documents) and the database gets deleted and reused each time, as in the unit test setup.  I'll try to get some logs.

**Comment by borrrden** [comment]
I've found the mysterious force that deletes the files.  It is sqlite itself making some quite bold assumptions....

If you have open sqlite3 statements and you allow a "zombie close" via `sqlite3_close_v2` then when you get around to calling `sqlite3_finalize` on any of those statements, it will go into the pattern that @snej mentioned above "When the finalizer runs (on some background thread) and finalizes the statement, the SQLite connection finally [sic] closes the db"

However, this involves such things as attempting to merge the WAL with the DB, deleting the SHM file, etc.  If a new sqlite3 db is opened on the same path then when the old zombie one comes along to close it, it thinks it needs to delete the WAL and SHM files but they are no longer owned by the zombie so it is deleting, essentially, another db's files.

**Comment by snej** [comment]
I've been musing about that possibility today: the zombie connection eating the brains of the new database file... I should fire off a question to the SQLite mailing list.

Avoiding this zombie mode would be nontrivial, though. We started using it because Java (and C#?) use finalizers that can be significantly delayed, and we don't want c4db_close to fail because finalizers haven't been run yet.

**Comment by borrrden** [comment]
The result of this is that there is a per-sqlite handle setting that basically says "Don't eat the WAL and SHM for dinner when you close" so after detecting the zombie state (i.e. failure of `sqlite3_close`, and the fallback to `sqlite3_close_v2`) we can set this setting to make sure the zombie doesn't chow down on the living WAL / SHM.

**Comment by snej** [comment]
Assigning back to QA to verify the fix -- you'll need to get an updated Android build with this commit cherry-picked into it, so coordinate with @hideki on that.

**Comment by hideki** [comment]
@djpongh 
Do we need this fix for the beta release?

**Comment by djpongh** [comment]
No fix for beta - we'll release note it.
